### PR TITLE
Added ApacheCommonsFileUtils WriteStringToFile Refaster templates

### PR DESCRIPTION
**Pull request by AlekSimpson**
## What's changed?
This PR adds two new refaster templates for the `FileUtils` migration. 

## What's your motivation?
There are still many `FileUtils` methods that don't have refaster templates setup yet.

## Anything in particular you'd like reviewers to focus on?
Both `CopyDirectory` and `WriteStringToFile` throw `java: reference to compile is ambiguous` errors. I think this could maybe be an issue with the templating engine but I'm not sure. 

## Anyone you would like to review specifically?
@timtebeek 

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've added the license header to any new files through `./gradlew licenseFormat`
- [x] I've used the IntelliJ IDEA auto-formatter on affected files
- [ ] I've updated the documentation (if applicable)

### CODE REVIEW

**Comment by timtebeek** [comment]
Sure; we'll try to work the Refaster template processor; I imagine you know what changes you'd need in the tests here?

**Comment by AlekSimpson** [comment]
> Sure; we'll try to work the Refaster template processor; I imagine you know what changes you'd need in the tests here?

Ya currently I think we have both `copyDirectory` and `writeStringToFile` in the tests. When I am able to build these templates I'll check and make sure it makes sense with the tests.

File: src/main/java/org/openrewrite/java/migrate/apache/commons/io/ApacheCommonsFileUtils.java [hunk 1]
```diff
@@ -47,5 +48,30 @@ File after(String name) {
...
 
+    private static class CopyDirectory {
+        @BeforeTemplate
+        void before(File a, File b) throws Exception {
+            FileUtils.copyDirectory(a, b);
+        }
+
+        @AfterTemplate
+        void after(File a, File b) throws Exception {
+            Files.copy(a.toPath(), b.toPath());
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/java/migrate/apache/commons/io/ApacheCommonsFileUtils.java:59]: I think the copy options `REPALCE_EXISTING` and probably also `COPY_ATTRIBUTES` should be specified to maintain a behavior as close as possible to that of `FileUtils`. However, in many cases users may not actually be really dependent on `COPY_ATTRIBUTES`, so I am unsure what to do about that one. Also, this is still just an approximation of what `FileUtils` does.

File: src/main/java/org/openrewrite/java/migrate/apache/commons/io/ApacheCommonsFileUtils.java [hunk 2]
```diff
@@ -47,5 +48,30 @@ File after(String name) {
...
+    @SuppressWarnings("deprecation")
+    private static class WriteStringToFile {
+        @BeforeTemplate
+        void before(File a, String s) throws Exception {
+            FileUtils.writeStringToFile(a, s);
+        }
+
+        @AfterTemplate
+        void after(File a, String s) throws Exception {
+            Files.write(a.toPath(), s.getBytes());
```
Comment (knutwannheden) [inline src/main/java/org/openrewrite/java/migrate/apache/commons/io/ApacheCommonsFileUtils.java:72]: This looks good, AFAICT. We could also add additional recipes to cover some of the other signatures (like `writeStringToFile(File, String, boolean)`, `writeStringToFile(File, String, Charset)`, and `writeStringToFile(File, String, Charset, boolean)`). But it is probably worth doing a search in the SaaS to check how often these get used. Probably they could all be covered by the same recipe, just using different before templates (all with the same input signatures).

Comment (AlekSimpson) [reply src/main/java/org/openrewrite/java/migrate/apache/commons/io/ApacheCommonsFileUtils.java:72]: Sounds good, I'll check the SaaS and see how often these overloads get used. I wanted to make sure this first template was accurate before worrying about overloads.

**Comment by timtebeek** [comment]
Found some commits I hadn't pushed earlier; not sure if our changes were merged correctly. 🤔

File: src/main/java/org/openrewrite/java/migrate/apache/commons/io/ApacheCommonsFileUtils.java (hunk 1 above)
Comment (timtebeek) [reply src/main/java/org/openrewrite/java/migrate/apache/commons/io/ApacheCommonsFileUtils.java:59]: If this is not a direct replacement, maybe best to leave it out for now @AlekSimpson

File: src/main/java/org/openrewrite/java/migrate/apache/commons/io/ApacheCommonsFileUtils.java (hunk 2 above)
Comment (timtebeek) [reply src/main/java/org/openrewrite/java/migrate/apache/commons/io/ApacheCommonsFileUtils.java:72]: Sounds like we could use a few more overloads here and merge those separate from the above `CopyDirectory`. Would you mind finishing this one @AlekSimpson ?

File: src/main/java/org/openrewrite/java/migrate/apache/commons/io/ApacheCommonsFileUtils.java [hunk 3]
```diff
@@ -47,5 +48,30 @@ File after(String name) {
...
+        void before(File a, File b) throws Exception {
+            FileUtils.copyDirectory(a, b);
+        }
+
+        @AfterTemplate
+        void after(File a, File b) throws Exception {
+            Files.copy(a.toPath(), b.toPath());
+        }
+    }
+
```
Comment (timtebeek) [inline src/main/java/org/openrewrite/java/migrate/apache/commons/io/ApacheCommonsFileUtils.java:62]:
```suggestion
```

File: src/main/java/org/openrewrite/java/migrate/apache/commons/io/ApacheCommonsFileUtils.java (hunk 2 above)
Comment (timtebeek) [reply src/main/java/org/openrewrite/java/migrate/apache/commons/io/ApacheCommonsFileUtils.java:72]: Would also be good to add a single document example test that shows which variants are replaced.

**Comment by AlekSimpson** [comment]
@timtebeek I'm now remembering more context around this PR from when I took a break with it. It seems the replacement method for `writeStringToFile` returns a `Path` but `writeStringToFile` returns void. Was this something you and @knutwannheden were planning on fixing with the refaster engine or should I try and find a write method that also returns void? If I remember correctly though, `Files.write()` is the only method that the Java `Files` class provides for writing to files.

**Comment by timtebeek** [comment]
ah yes; that's where we got stuck before; would it help to add explicit casts to the `after` template that matches the first `before` template type @knutwannheden ? I think we only replace like-for-like here right? Given that the context can differ..

**Comment by knutwannheden** [comment]
I will try to take a look. Should be something we can fix.

**Comment by knutwannheden** [comment]
Should now be fixed with https://github.com/openrewrite/rewrite-templating/commit/b472f3ad4723b36cfb9cfae8a265ab981f1a1b6d.

**Review by timtebeek — APPROVED** [review]
Let's do the overloads in a separate PR. Thanks for the help @knutwannheden

**Pull request merged**